pkg/network/protocols/events/README.md

```diff
@@ -0,0 +1,73 @@
+# `pkg/network/protocols/events`
+
+## Motivation
+
+The purpose of this package is to standardize the communication between user and
+kernel space across different Universal Service Monitoring (USM) protocols.
+
+Some technical concerns that were factored in the code and are worth mentioning are:
+
+* eBPF programs of type `BPF_PROG_TYPE_SOCKET_FILTER` are not (for _most_ Kernel
+  versions) able to directly write events into perf buffers;
+* Reading events off-perf buffers can be quite CPU-intensive. This can be
+mitigated by properly configuring the `watermark` option when issuing the
+`perf_event_open` syscall. However, there is currently no way to _synchronize_
+(eg. flush all buffered events) at a given moment in time[^1], which is a
+requirement for us.
+
+[^1]: this may be available in the near future since we could probably force
+wake-up events on the fly via `ioctl` calls, but this will likely require us to
+upstream changes to the `cilium/ebpf` library. There is a Jira card owned by
+`ebpf-platform` tracking this, and once this is available we could _greatly_
+simplify this package.
+
+
+## How to use the library
+
+### Kernel Side
+
+```c
+#include "protocols/events.h"
+USM_EVENTS_INIT(<protocol>, <event_type>, <batch_size>);
+```
+
+This will instantiate the necessary eBPF maps along with two functions:
+* `<protocol>_batch_enqueue`;
+* `<protocol>_batch_flush`;
+
+Please note that `<proto>_batch_flush` requires access to the
+`bpf_perf_event_output` helper, which is typically not available to socket
+filter programs. Because of that we recommend to call it from
+`netif_receive_skb` which is associated to the execution of socket filter programs:
+
+```c
+SEC("tracepoint/net/netif_receive_skb")
+int tracepoint__net__netif_receive_skb(struct pt_regs* ctx) {
+    <protocol>_flush_batch(ctx);
+    return 0;
+}
+```
```

Comment:
Yes. The fixed cost would roughly be one eBPF map lookup per protocol.

Reply:
consider change the example so it will be clear for people they need to add a new line here

Reply:
that's a good point, but now that I think about it, I think it may be misleading to add it because it assumes that all protocols will be bundled within the same compilation artifact